File: code/backend/src/llm_geometry/geo/fields.py
"""The two Geometry Lab vector-field modes, exactly per the frozen contract.

`next_next` (the issue's field): for every vocab token *v* hypothetically appended to
the prompt, arrows from ``points[v]`` toward the embedding(s) of the following-token
prediction. ``layer`` selects which layer's residual stream feeds the (tied, logit-lens)
readout; ``"full"`` = the final layer. ``temperature=0 & top_m=1`` ⇒ one argmax arrow
per point with weight 1; otherwise the ``top_m`` most probable targets, each arrow
weighted by its raw probability. At ``temperature=0`` the distribution is one-hot, so
only the argmax arrow (the only one with positive weight) is emitted regardless of
``top_m``.

`force` (arXiv:2607.13295): the per-point field ``W_V·z`` over all vocab points for
the chosen layer (``antisymmetrize=true`` uses ``(W_V−W_Vᵀ)/2``, whose field is
*exactly* tangent to the sphere: ⟨Az, z⟩ = 0 for antisymmetric A), plus per-sequence-
position aggregate forces ``Σ_{j≤i} softmax(⟨K z_j, Q z_i⟩)·V z_j`` — literally the
model's ``attention @ v`` rows.

Those aggregate forces are **projected onto the tangent plane at their anchor point**
``z_i`` before being returned, and the magnitude of the removed radial component is
reported as ``normal_residual``. This matters: the arrow is drawn anchored at ``z_i``
on the sphere, and antisymmetrizing ``W_V`` does NOT make the sum tangent there —
each term ``W_V z_j`` is tangent at ``z_j``, not at ``z_i``. Projecting is a display
choice, and ``normal_residual`` is how the UI stays honest about what it hides.

The antisymmetrize toggle applies to the per-point field only; the aggregate forces
always use the real W_V. ``layer="full"`` is invalid for force mode (it is per-layer
by definition).

Also home to the spherical-binning helpers shared with the training gate metrics
(Fibonacci-lattice bins; directional entropy; coverage uniformity).
"""

from __future__ import annotations

from typing import Any

import numpy as np
import torch

from ..errors import InvalidParamError
from .config import CONTEXT_WINDOW, N_LAYERS, SPHERE_BINS, VOCAB_SIZE
from .model import GeoTransformer, resolve_layer


def _clip_prompt(prompt_ids: list[int], room: int) -> list[int]:
    """Keep the most recent tokens, leaving ``room`` positions free (LM conditioning)."""
    max_len = CONTEXT_WINDOW - room
    return list(prompt_ids)[-max_len:] if len(prompt_ids) > max_len else list(prompt_ids)


@torch.no_grad()
def next_next_field(
    model: GeoTransformer,
    prompt_ids: list[int],
    layer: int | str = "full",
    temperature: float = 0.0,
    top_m: int = 1,
) -> dict[str, Any]:
    if temperature is None or float(temperature) < 0:
        raise InvalidParamError(f"temperature must be >= 0, got {temperature!r}")
    if int(top_m) < 1:
        raise InvalidParamError(f"top_m must be >= 1, got {top_m!r}")
    layer_idx = resolve_layer(layer)

    prompt = _clip_prompt([int(t) for t in prompt_ids], room=1)
    points = model.embedding.detach().cpu().numpy().astype(np.float32)  # (V, 3)

    # One batched forward: every vocab token appended to the same prompt.
    vocab = torch.arange(VOCAB_SIZE, dtype=torch.long).unsqueeze(1)  # (V, 1)
    if prompt:
        prefix = torch.tensor(prompt, dtype=torch.long).unsqueeze(0).expand(VOCAB_SIZE, -1)
        ids = torch.cat([prefix, vocab], dim=1)  # (V, P+1)
    else:
        ids = vocab
    run = model._run(ids, need_trace=False)
    h_last = run["hidden_out"][layer_idx][:, -1, :]  # (V, 3)
    logits = model.readout(h_last)  # (V, V)

    arrows: list[dict[str, Any]] = []
    if float(temperature) == 0.0:
        targets = torch.argmax(logits, dim=-1).cpu().numpy()  # (V,)
        for v in range(VOCAB_SIZE):
            vec = points[int(targets[v])] - points[v]
            arrows.append({"origin_index": v, "vec": vec.tolist(), "weight": 1.0})
    else:
        probs = torch.softmax(logits / float(temperature), dim=-1)
        m = min(int(top_m), VOCAB_SIZE)
        top_p, top_i = torch.topk(probs, m, dim=-1)
        top_p_np = top_p.cpu().numpy()
        top_i_np = top_i.cpu().numpy()
        for v in range(VOCAB_SIZE):
            for rank in range(m):
                weight = float(top_p_np[v, rank])
                if weight <= 0.0:
                    continue
                vec = points[int(top_i_np[v, rank])] - points[v]
                arrows.append({"origin_index": v, "vec": vec.tolist(), "weight": weight})

    return {
        "mode": "next_next",
        "layer": "full" if layer == "full" else layer_idx,
        "points": points,
        "token_ids": list(range(VOCAB_SIZE)),
        "arrows": arrows,
        "sequence_forces": None,
        "tangent_exact": False,
    }


@torch.no_grad()
def force_field(
    model: GeoTransformer,
    prompt_ids: list[int],
    layer: int | str,
    antisymmetrize: bool = False,
) -> dict[str, Any]:
    if layer == "full":
        raise InvalidParamError(
            'layer="full" is invalid for force mode: the attention force is per-layer '
            "by definition; choose a layer 0..3"
        )
    layer_idx = resolve_layer(layer)
    if not 0 <= layer_idx < N_LAYERS:  # unreachable, resolve_layer guards; keep explicit
        raise InvalidParamError(f"layer must be in 0..{N_LAYERS - 1}")

    points = model.embedding.detach().cpu().numpy().astype(np.float32)  # (V, 3)
    w_v = model.layers[layer_idx].W_V.detach().cpu().numpy().astype(np.float32)
    w_eff = 0.5 * (w_v - w_v.T) if antisymmetrize else w_v

    vecs = points @ w_eff.T  # per-point field W_V·z, (V, 3)
    mags = np.linalg.norm(vecs, axis=1)
    max_mag = float(mags.max())
    weights = mags / max_mag if max_mag > 0 else np.zeros_like(mags)
    arrows = [
        {"origin_index": v, "vec": vecs[v].tolist(), "weight": float(weights[v])}
        for v in range(VOCAB_SIZE)
    ]

    sequence_forces: list[dict[str, Any]] = []
    prompt = _clip_prompt([int(t) for t in prompt_ids], room=0)
    if prompt:
        ids = torch.tensor(prompt, dtype=torch.long).unsqueeze(0)
        run = model._run(ids, need_trace=True)
        tr = run["trace"][layer_idx]
        attn = tr["attention"][0]  # (T, T) row-stochastic, causal
        v_proj = tr["v"][0]  # (T, 3) — the real W_V z_j
        forces = (attn @ v_proj).cpu().numpy().astype(np.float32)  # Σ_{j≤i} A_ij V z_j
        # Project at the point the arrow is actually DRAWN at — the prompt token's
        # embedding, which is the unit-norm point on the sphere the client anchors to
        # (`GeoTrace.embeddings`). Projecting at the layer's residual stream
        # (`hidden_in`) instead was wrong: the residual stream is not on the sphere and
        # drifts away from the token embedding with depth, so the "tangent" arrows came
        # out up to 59° off the tangent plane at layer 2 while the UI claimed otherwise.
        #
        # Antisymmetrizing W_V does not help here either: each term W_V z_j is tangent
        # at z_j, not at the anchor where the sum is drawn. `normal_residual` reports the
        # radial magnitude removed, so nothing is hidden.
        anchors = points[np.asarray(prompt, dtype=np.int64)]  # (T, 3) unit-norm
        anchor_norms = np.linalg.norm(anchors, axis=1)
        for i in range(forces.shape[0]):
            if anchor_norms[i] > 1e-12:
                n_hat = anchors[i] / anchor_norms[i]
                radial = float(forces[i] @ n_hat)
                tangential = forces[i] - radial * n_hat
            else:
                radial = 0.0
                tangential = forces[i]
            sequence_forces.append(
                {
                    "position": i,
                    "vec": tangential.tolist(),
                    "normal_residual": abs(radial),
                }
            )

    return {
        "mode": "force",
        "layer": layer_idx,
        "points": points,
        "token_ids": list(range(VOCAB_SIZE)),
        "arrows": arrows,
        "sequence_forces": sequence_forces,
        "tangent_exact": bool(antisymmetrize),
    }


# -- spherical binning + gate metrics --------------------------------------------------


def fibonacci_sphere(n: int = SPHERE_BINS) -> np.ndarray:
    """(n, 3) unit vectors on a golden-angle Fibonacci lattice (near-equal-area bins)."""
    i = np.arange(n, dtype=np.float64) + 0.5
    phi = np.arccos(1.0 - 2.0 * i / n)
    theta = np.pi * (1.0 + np.sqrt(5.0)) * i
    return np.stack(
        [np.sin(phi) * np.cos(theta), np.sin(phi) * np.sin(theta), np.cos(phi)], axis=1
    ).astype(np.float32)


def _bin_entropy(unit_vectors: np.ndarray, n_bins: int) -> float:
    """Shannon entropy (nats) of nearest-bin occupancy for unit vectors on S²."""
    if unit_vectors.shape[0] == 0:
        return 0.0
    bins = fibonacci_sphere(n_bins)
    assignment = np.argmax(unit_vectors @ bins.T, axis=1)
    counts = np.bincount(assignment, minlength=n_bins).astype(np.float64)
    p = counts / counts.sum()
    p = p[p > 0]
    return float(-(p * np.log(p)).sum())


def coverage_uniformity(points: np.ndarray, n_bins: int = SPHERE_BINS) -> float:
    """Dispersion of embedding points over the sphere, normalized to 0..1."""
    norms = np.linalg.norm(points, axis=1, keepdims=True)
    unit = points / np.maximum(norms, 1e-12)
    return _bin_entropy(unit.astype(np.float32), n_bins) / float(np.log(n_bins))


def field_directional_entropy(arrows: list[dict[str, Any]], n_bins: int = SPHERE_BINS) -> float:
    """Entropy (nats) of arrow *directions* binned on the sphere. Zero-length arrows
    (self-predictions) carry no direction and are excluded."""
    vecs = np.asarray([a["vec"] for a in arrows], dtype=np.float64)
    if vecs.size == 0:
        return 0.0
    mags = np.linalg.norm(vecs, axis=1)
    keep = mags > 1e-8
    if not np.any(keep):
        return 0.0
    unit = (vecs[keep] / mags[keep, None]).astype(np.float32)
    return _bin_entropy(unit, n_bins)

File: code/frontend/src/lib/geoEngine/fields.ts
/**
 * The two Geometry Lab vector-field modes — TypeScript port of
 * code/backend/src/llm_geometry/geo/fields.py, returning objects shaped exactly
 * like the frozen /api/geo/vector_field contract (numbers unrounded; the backend's
 * 6-significant-digit JSON rounding is a transport concern, not applied here).
 *
 * next_next: for every vocab token v hypothetically appended to the prompt, arrows
 * from points[v] toward the embedding(s) of the following-token prediction. The
 * backend runs one batched forward over all 1003 (prompt + v) sequences; because
 * the prompt prefix is identical across the batch and attention is causal, the
 * prefix hidden states / keys / values are identical too — we compute them once
 * and run only the appended position per vocab token (same math, ~50x faster).
 *
 * force: per-point field W_V·z over all vocab points (antisymmetrize uses
 * (W_V - W_V^T)/2, exactly tangent to the sphere), plus per-sequence-position
 * aggregate forces sum_{j<=i} softmax(<K z_j, Q z_i>) V z_j — literally the
 * model's attention @ v rows — with normal residual magnitudes |<f_i, z_hat_i>|.
 * The antisymmetrize toggle applies to the per-point field ONLY; aggregate forces
 * always use the real W_V. layer="full" is invalid for force mode.
 */

import { invalidParam } from "./errors";
import {
  CONTEXT_WINDOW,
  D_MODEL,
  GeoModel,
  MLP_HIDDEN,
  N_LAYERS,
  VOCAB_SIZE,
  resolveLayer,
  type LayerParam,
} from "./model";
import { argmax, argsortDesc, dot, gelu, softmaxInPlace, toNested2 } from "./tensor";

export interface FieldArrow {
  origin_index: number;
  vec: number[];
  weight: number;
}

export interface SequenceForce {
  position: number;
  vec: number[];
  normal_residual: number;
}

export interface VectorFieldResult {
  mode: "next_next" | "force";
  layer: LayerParam;
  points: number[][];
  token_ids: number[];
  arrows: FieldArrow[];
  sequence_forces: SequenceForce[] | null;
  tangent_exact: boolean;
}

/** Keep the most recent tokens, leaving `room` positions free (LM conditioning). */
export function clipPrompt(promptIds: number[], room: number): number[] {
  const maxLen = CONTEXT_WINDOW - room;
  return promptIds.length > maxLen ? promptIds.slice(-maxLen) : [...promptIds];
}

const allTokenIds = (): number[] => Array.from({ length: VOCAB_SIZE }, (_, i) => i);

interface PrefixLayer {
  k: Float64Array; // (P,3) prefix keys at this layer
  v: Float64Array; // (P,3) prefix values at this layer
  T: number; // P
}

/**
 * Hidden state of the LAST position (the appended vocab token) after layers
 * 0..upTo-1, given cached prefix keys/values per layer. Exactly the math of the
 * backend's batched _run restricted to the final causal row.
 */
function appendedHidden(
  model: GeoModel,
  prefix: PrefixLayer[],
  tokenId: number,
  position: number,
  upTo: number,
): Float64Array {
  const E = model.embedding;
  const pos = model.ws["pos_embedding"];
  const h = new Float64Array(D_MODEL);
  for (let c = 0; c < D_MODEL; c++) h[c] = E[tokenId * D_MODEL + c] + pos[position * D_MODEL + c];

  for (let l = 0; l < upTo; l++) {
    const WQ = model.layerParam(l, "W_Q");
    const WK = model.layerParam(l, "W_K");
    const WV = model.layerParam(l, "W_V");
    const WO = model.layerParam(l, "W_O");
    const q = new Float64Array(D_MODEL);
    const k = new Float64Array(D_MODEL);
    const v = new Float64Array(D_MODEL);
    for (let a = 0; a < D_MODEL; a++) {
      for (let b = 0; b < D_MODEL; b++) {
        q[a] += WQ[a * D_MODEL + b] * h[b];
        k[a] += WK[a * D_MODEL + b] * h[b];
        v[a] += WV[a * D_MODEL + b] * h[b];
      }
    }
    const P = prefix[l].T;
    const pk = prefix[l].k;
    const pv = prefix[l].v;
    const scores = new Float64Array(P + 1);
    for (let j = 0; j < P; j++) scores[j] = dot(q, pk, D_MODEL, 0, j * D_MODEL);
    scores[P] = dot(q, k, D_MODEL);
    softmaxInPlace(scores);
    const ctx = new Float64Array(D_MODEL);
    for (let j = 0; j < P; j++) {
      const a = scores[j];
      for (let c = 0; c < D_MODEL; c++) ctx[c] += a * pv[j * D_MODEL + c];
    }
    for (let c = 0; c < D_MODEL; c++) ctx[c] += scores[P] * v[c];
    for (let a = 0; a < D_MODEL; a++) {
      let s = 0;
      for (let b = 0; b < D_MODEL; b++) s += WO[a * D_MODEL + b] * ctx[b];
      h[a] += s;
    }
    // MLP block: h += gelu(h W_in + b_in) W_out + b_out
    const WIn = model.layerParam(l, "W_in");
    const bIn = model.layerParam(l, "b_in");
    const WOut = model.layerParam(l, "W_out");
    const bOut = model.layerParam(l, "b_out");
    const mlpOut = new Float64Array(D_MODEL);
    for (let j = 0; j < MLP_HIDDEN; j++) {
      let pre = bIn[j];
      for (let b = 0; b < D_MODEL; b++) pre += h[b] * WIn[b * MLP_HIDDEN + j];
      const g = gelu(pre);
      for (let c = 0; c < D_MODEL; c++) mlpOut[c] += g * WOut[j * D_MODEL + c];
    }
    for (let c = 0; c < D_MODEL; c++) h[c] += mlpOut[c] + bOut[c];
  }
  return h;
}

export function nextNextField(
  model: GeoModel,
  promptIds: number[],
  layer: LayerParam = "full",
  temperature = 0,
  topM = 1,
): VectorFieldResult {
  if (temperature === null || temperature === undefined || Number(temperature) < 0) {
    throw invalidParam(`temperature must be >= 0, got ${JSON.stringify(temperature)}`);
  }
  if (Math.trunc(Number(topM)) < 1) {
    throw invalidParam(`top_m must be >= 1, got ${JSON.stringify(topM)}`);
  }
  const layerIdx = resolveLayer(layer);
  const upTo = layerIdx + 1;

  const prompt = clipPrompt(promptIds.map((t) => Math.trunc(t)), 1);
  const E = model.embedding;
  const points = toNested2(E, VOCAB_SIZE, D_MODEL);

  // Prefix cache: run the prompt once, keep per-layer keys/values (identical for
  // every appended vocab token by causality).
  let prefix: PrefixLayer[];
  if (prompt.length > 0) {
    const acts = model.forwardSeq(prompt, upTo);
    prefix = acts.layers.map((la) => ({ k: la.k, v: la.v, T: acts.T }));
  } else {
    prefix = Array.from({ length: upTo }, () => ({
      k: new Float64Array(0),
      v: new Float64Array(0),
      T: 0,
    }));
  }
  const position = prompt.length; // the appended token's position index

  const arrows: FieldArrow[] = [];
  const T0 = Number(temperature) === 0;
  const m = Math.min(Math.trunc(Number(topM)), VOCAB_SIZE);

  for (let vTok = 0; vTok < VOCAB_SIZE; vTok++) {
    const h = appendedHidden(model, prefix, vTok, position, upTo);
    const logits = model.readoutOne(h);
    if (T0) {
      const target = argmax(logits);
      arrows.push({
        origin_index: vTok,
        vec: [
          E[target * D_MODEL] - E[vTok * D_MODEL],
          E[target * D_MODEL + 1] - E[vTok * D_MODEL + 1],
          E[target * D_MODEL + 2] - E[vTok * D_MODEL + 2],
        ],
        weight: 1,
      });
    } else {
      const t = Number(temperature);
      for (let i = 0; i < logits.length; i++) logits[i] /= t;
      softmaxInPlace(logits);
      const order = argsortDesc(logits);
      for (let rank = 0; rank < m; rank++) {
        const target = order[rank];
        // Quantize to float32 like the backend's f32 softmax before the >0 check,
        // so sub-float32 tail probabilities are dropped identically.
        const weight = Math.fround(logits[target]);
        if (weight <= 0) continue;
        arrows.push({
          origin_index: vTok,
          vec: [
            E[target * D_MODEL] - E[vTok * D_MODEL],
            E[target * D_MODEL + 1] - E[vTok * D_MODEL + 1],
            E[target * D_MODEL + 2] - E[vTok * D_MODEL + 2],
          ],
          weight,
        });
      }
    }
  }

  return {
    mode: "next_next",
    layer: layer === "full" ? "full" : layerIdx,
    points,
    token_ids: allTokenIds(),
    arrows,
    sequence_forces: null,
    tangent_exact: false,
  };
}

export function forceField(
  model: GeoModel,
  promptIds: number[],
  layer: LayerParam,
  antisymmetrize = false,
): VectorFieldResult {
  if (layer === "full") {
    throw invalidParam(
      'layer="full" is invalid for force mode: the attention force is per-layer by definition; ' +
        "choose a layer 0..3",
    );
  }
  const layerIdx = resolveLayer(layer);
  if (layerIdx < 0 || layerIdx >= N_LAYERS) {
    throw invalidParam(`layer must be in 0..${N_LAYERS - 1}`);
  }

  const E = model.embedding;
  const points = toNested2(E, VOCAB_SIZE, D_MODEL);
  const wV = model.layerParam(layerIdx, "W_V");
  const wEff = new Float64Array(9);
  if (antisymmetrize) {
    for (let a = 0; a < 3; a++) {
      for (let b = 0; b < 3; b++) wEff[a * 3 + b] = 0.5 * (wV[a * 3 + b] - wV[b * 3 + a]);
    }
  } else {
    for (let i = 0; i < 9; i++) wEff[i] = wV[i];
  }

  // Per-point field W_eff · z over all vocab points; weights = |vec| / max |vec|.
  const vecs = new Float64Array(VOCAB_SIZE * 3);
  const mags = new Float64Array(VOCAB_SIZE);
  let maxMag = 0;
  for (let vTok = 0; vTok < VOCAB_SIZE; vTok++) {
    for (let a = 0; a < 3; a++) {
      let s = 0;
      for (let b = 0; b < 3; b++) s += wEff[a * 3 + b] * E[vTok * 3 + b];
      vecs[vTok * 3 + a] = s;
    }
    const mag = Math.hypot(vecs[vTok * 3], vecs[vTok * 3 + 1], vecs[vTok * 3 + 2]);
    mags[vTok] = mag;
    if (mag > maxMag) maxMag = mag;
  }
  const arrows: FieldArrow[] = [];
  for (let vTok = 0; vTok < VOCAB_SIZE; vTok++) {
    arrows.push({
      origin_index: vTok,
      vec: [vecs[vTok * 3], vecs[vTok * 3 + 1], vecs[vTok * 3 + 2]],
      weight: maxMag > 0 ? mags[vTok] / maxMag : 0,
    });
  }

  // Per-sequence-position aggregate forces (always the REAL W_V, never antisymmetrized).
  //
  // Mirrors geo/fields.py: the sum is drawn anchored at the prompt token's EMBEDDING,
  // so it is projected onto the tangent plane there and the removed radial magnitude is
  // reported as normal_residual. Antisymmetrizing W_V would not help — each term
  // W_V z_j is tangent at z_j, not at the anchor.
  const sequenceForces: SequenceForce[] = [];
  const prompt = clipPrompt(promptIds.map((t) => Math.trunc(t)), 0);
  if (prompt.length > 0) {
    const acts = model.forwardSeq(prompt, layerIdx + 1);
    const tr = acts.layers[layerIdx];
    const T = acts.T;
    for (let i = 0; i < T; i++) {
      const force = [0, 0, 0];
      for (let j = 0; j <= i; j++) {
        const a = tr.attention[i * T + j];
        for (let c = 0; c < 3; c++) force[c] += a * tr.v[j * 3 + c];
      }
      // Project at the point the arrow is DRAWN at — the prompt token's embedding,
      // which is the unit-norm sphere point the client anchors to. Using the layer's
      // residual stream (hiddenIn) put the "tangent" arrows up to 59° out of plane.
      const a = prompt[i] * 3;
      const z0 = model.embedding[a];
      const z1 = model.embedding[a + 1];
      const z2 = model.embedding[a + 2];
      const zNorm = Math.hypot(z0, z1, z2);
      let radial = 0;
      if (zNorm > 1e-12) {
        const n0 = z0 / zNorm;
        const n1 = z1 / zNorm;
        const n2 = z2 / zNorm;
        radial = force[0] * n0 + force[1] * n1 + force[2] * n2;
        force[0] -= radial * n0;
        force[1] -= radial * n1;
        force[2] -= radial * n2;
      }
      sequenceForces.push({ position: i, vec: force, normal_residual: Math.abs(radial) });
    }
  }

  return {
    mode: "force",
    layer: layerIdx,
    points,
    token_ids: allTokenIds(),
    arrows,
    sequence_forces: sequenceForces,
    tangent_exact: Boolean(antisymmetrize),
  };
}
